### DISCUSSION TITLE
Server components and data duplication

### DISCUSSION THREAD

**Discussion by Kein1945** [vercel/next.js#56283] [Help] [answered]
### Summary

I just want to clarify to understand: data which is used to render on server side is duplicated inside HTML, am I correct?

I have an idea to migrate our .Net website to Next.js along with a backend API. We've been focusing on performance optimization, and due to the split between new React and old jQuery/Bootstrap code, we're facing size-related issues. My thought was to transition fully to React and Next.js while keeping .Net solely for the API. However, my main concern is that with server-side rendering (SSR), we still need to pass component data as window.__INITIAL_STATE__, resulting in data duplication. I was hopeful that Server Components might address this issue, especially for our large text-heavy pages. However, I noticed that even a simple example like the one below duplicates API data in the page source:

```jsx
async function getData() {
  const res = await fetch('https://jsonplaceholder.typicode.com/posts/1')
  if (!res.ok) {
    throw new Error('Failed to fetch data')
  }
 
  return res.json()
}

export default async function Home() {
  const data = await getData();
  return data.body;
}
```
So my question is: Am I doing something wrong, or is this behavior expected? If it's expected, is there a way to mitigate this for performance purposes, or is Next.js not a suitable choice in this case?

p.s. I understand why I need window.__INITIAL_STATE__ for SSR, because of react hydration, just want to clarify what is Server components and what is not.

**Comment by icyJoseph** [comment]
Mmm while I explore this a bit, could you also take a look at this video? He even uses a tool to measure how much "data" you end up sending to the client: 

- https://www.youtube.com/watch?v=FfHsIio4aCU

**Reply by icyJoseph** [reply]
I see, the text node that gets returned from `data.body` ends up as part of the RSC. So in this case, then yeah, it duplicates that data. But if you fetched N items, and from those select M properties as actual data, then the RSC should only have those M properties. 

However, I guess, the data you "work" with ends up as part of the RSC, to build the Client Tree integration.

**Comment by christiankaindl** [comment, marked as answer]
Sadly, this seems to be a current limitation of Next.js' app router (or rather React?). See this comment from an old ticket (one of the checkboxes that is not ticked): https://github.com/vercel/next-react-server-components/issues/13#issuecomment-988438962

> As an additional optimization on top of this inlining, we'll want to deduplicate text content between them. E.g. we shouldn't need to repeat all the story titles in the Flight data structure even though they have just appeared in the HTML itself. We need to add a capability in React to "skip" specially marked Flight text and keep using the one already in the DOM, thereby letting us deduplicate it.

So the Server Components architecture would make it theoretically possible to skip data duplication (so that data doesn't end up in the HTML _and_ the RSC payload), but this optimization is not yet implemented.

Also, see this discussion (you might want to upvote it!): https://github.com/vercel/next.js/discussions/42170

**Reply by Kein1945** [reply]
I should found that question(https://github.com/vercel/next.js/discussions/42170), thank you for help, @christiankaindl !
